### PR TITLE
Staking pool factory

### PR DESCRIPTION
**Pull request by evgenykuzyakov**
This contract allows to deploy new staking pool contract and automatically whitelist them.

- [x] Add a few more integration tests

### CODE REVIEW

File: staking-pool-factory/README.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
+# Staking Pool Factory Contract
+
+This contract allows to deploy new staking pool contract and automatically whitelist them.
```
Comment (willemneal) [inline staking-pool-factory/README.md:3]:
```suggestion
This contract deploys and automatically whitelists new staking pool contracts.
```

File: staking-pool-factory/README.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+It allows any user to create an new whitelisted staking pool.
+
+The staking pool factory contract keeps the hardcoded binary of the staking pool contract within itself.
```
Comment (willemneal) [inline staking-pool-factory/README.md:6]:
```suggestion
The staking pool factory contract packages the binary of the staking pool contract within its own binary.
```

File: staking-pool-factory/README.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+To create a new staking pool a user should issue a function call transaction and attach the required minimum deposit.
+The entire deposit will be transferred to the newly created staking pool contract in order to cover the required storage.
+
+When a user issues a function call towards the factory to create a new staking pool the factory internally checks that
```
Comment (willemneal) [inline staking-pool-factory/README.md:10]: What does it mean to internally check?  Does this mean that it checks that the account ID doesn't exist as an account on chain?

File: staking-pool-factory/README.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+the staking pool account ID doesn't exists, validates arguments for the staking pool initialization and then issues a
```
Comment (willemneal) [inline staking-pool-factory/README.md:11]: nit
```suggestion
the staking pool account ID does not exists, validates arguments for the staking pool initialization and then issues a
```

File: staking-pool-factory/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,373 @@
...
+static ALLOC: wee_alloc::WeeAlloc = wee_alloc::WeeAlloc::INIT;
+
+#[near_bindgen]
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct StakingPoolFactory {
+    /// Account ID of the staking pool whitelist contract.
+    staking_pool_whitelist_account_id: AccountId,
+
+    /// The account ID of the staking pools created.
+    staking_pool_account_ids: Set<AccountId>,
```
Comment (willemneal) [inline staking-pool-factory/src/lib.rs:45]: Isn't this information stored in the whitelist contract?

**Comment by willemneal** [comment]
LGTM just small nits no README and one question about design.

File: staking-pool-factory/src/lib.rs (hunk 5 above)
Comment (evgenykuzyakov) [reply staking-pool-factory/src/lib.rs:45]: It's different, here it mostly to avoid creation of the staking pools with the same name.

Comment (willemneal) [reply staking-pool-factory/src/lib.rs:45]: Yeah, but can't you make a promise call to the whitelist to do that?

Comment (evgenykuzyakov) [reply staking-pool-factory/src/lib.rs:45]: That's extra call, but in theory nothing breaks if we don't maintain the `Set` internally. It just saves some gas if 2 people try to create the same pool. Then one of the creations is going to fail.

Comment (willemneal) [reply staking-pool-factory/src/lib.rs:45]: Okay makes sense to me.  Then everything LGTM.

**Pull request merged**